### PR TITLE
Rustup to rustc 1.41.0-nightly (e87a205c2 2019-11-27)

### PR DESCRIPTION
**Pull request by flip1995**
Rustups:

- rust-lang/rust#66671 (Ast address-of)
- rust-lang/rust#64856 (Scope format! temporaries)
 - http://github.com/rust-lang/rust/pull/66719

changelog: none

### CODE REVIEW

File: tests/ui/or_fun_call.stderr [hunk 1]
```diff
@@ -72,17 +72,11 @@ error: use of `unwrap_or` followed by a function call
 LL |     let _ = stringy.unwrap_or("".to_owned());
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `unwrap_or_else(|| "".to_owned())`
 
-error: use of `ok_or` followed by a function call
-  --> $DIR/or_fun_call.rs:74:17
-   |
-LL |     let _ = opt.ok_or(format!("{} world.", hello));
-   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `ok_or_else(|| format!("{} world.", hello))`
```
Comment (flip1995) [inline tests/ui/or_fun_call.stderr]: `format!` now expands to a block. While format is explicitly handled in `expect_fun_call`, it is not in `or_fun_call`.

**Comment by Centril** [comment]
> [rust-lang/rust#66671](https://github.com/rust-lang/rust/pull/66671) (Ast address-of)

cc @matthewjasper -- simply doing `_` on the `BorrowKind` might give wrong linting behaviors depending on the lint. You might want to be more conservative and match on `::Ref` initially.

**Comment by flip1995** [comment]
Ah thanks, good point!

File: clippy_lints/src/eval_order_dependence.rs [hunk 2]
```diff
@@ -328,7 +328,7 @@ impl<'a, 'tcx> Visitor<'tcx> for ReadVisitor<'a, 'tcx> {
             // ```
             //
             // TODO: fix this
-            ExprKind::AddrOf(_, _) => {
+            ExprKind::AddrOf(_, _, _) => {
```
Comment (matthewjasper) [inline clippy_lints/src/eval_order_dependence.rs]: This can stay

File: clippy_lints/src/functions.rs [hunk 3]
```diff
@@ -656,7 +656,7 @@ impl<'a, 'tcx> intravisit::Visitor<'tcx> for StaticMutVisitor<'a, 'tcx> {
                     tys.clear();
                 }
             },
-            Assign(ref target, _) | AssignOp(_, ref target, _) | AddrOf(hir::Mutability::Mutable, ref target) => {
+            Assign(ref target, _) | AssignOp(_, ref target, _) | AddrOf(_, hir::Mutability::Mutable, ref target) => {
```
Comment (matthewjasper) [inline clippy_lints/src/functions.rs]: This can stay

File: clippy_lints/src/mutable_debug_assertion.rs [hunk 4]
```diff
@@ -77,14 +77,14 @@ fn extract_call<'a, 'tcx>(cx: &'a LateContext<'a, 'tcx>, e: &'tcx Expr) -> Optio
                         if let ExprKind::Tup(ref conditions) = headerexpr.kind;
                         if conditions.len() == 2;
                         then {
-                            if let ExprKind::AddrOf(_, ref lhs) = conditions[0].kind {
+                            if let ExprKind::AddrOf(_, _, ref lhs) = conditions[0].kind {
```
Comment (matthewjasper) [inline clippy_lints/src/mutable_debug_assertion.rs]: Maybe the ones in this file?

File: clippy_lints/src/no_effect.rs [hunk 5]
```diff
@@ -58,7 +58,7 @@ fn has_no_effect(cx: &LateContext<'_, '_>, expr: &Expr) -> bool {
         | ExprKind::Type(ref inner, _)
         | ExprKind::Unary(_, ref inner)
         | ExprKind::Field(ref inner, _)
-        | ExprKind::AddrOf(_, ref inner)
+        | ExprKind::AddrOf(_, _, ref inner)
```
Comment (matthewjasper) [inline clippy_lints/src/no_effect.rs]: The ones in this file

File: clippy_lints/src/reference.rs [hunk 6]
```diff
@@ -37,7 +37,7 @@ impl EarlyLintPass for DerefAddrOf {
     fn check_expr(&mut self, cx: &EarlyContext<'_>, e: &Expr) {
         if_chain! {
             if let ExprKind::Unary(UnOp::Deref, ref deref_target) = e.kind;
-            if let ExprKind::AddrOf(_, ref addrof_target) = without_parens(deref_target).kind;
+            if let ExprKind::AddrOf(_, _, ref addrof_target) = without_parens(deref_target).kind;
```
Comment (matthewjasper) [inline clippy_lints/src/reference.rs]: The ones in this file

File: clippy_lints/src/shadow.rs [hunk 7]
```diff
@@ -313,9 +313,10 @@ fn check_expr<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr, bindings:
         return;
     }
     match expr.kind {
-        ExprKind::Unary(_, ref e) | ExprKind::Field(ref e, _) | ExprKind::AddrOf(_, ref e) | ExprKind::Box(ref e) => {
-            check_expr(cx, e, bindings)
-        },
+        ExprKind::Unary(_, ref e)
+        | ExprKind::Field(ref e, _)
+        | ExprKind::AddrOf(_, _, ref e)
```
Comment (matthewjasper) [inline clippy_lints/src/shadow.rs]: The ones in this file

File: clippy_lints/src/utils/hir_utils.rs [hunk 8]
```diff
@@ -78,7 +78,7 @@ impl<'a, 'tcx> SpanlessEq<'a, 'tcx> {
         }
 
         match (&left.kind, &right.kind) {
-            (&ExprKind::AddrOf(l_mut, ref le), &ExprKind::AddrOf(r_mut, ref re)) => {
+            (&ExprKind::AddrOf(_, l_mut, ref le), &ExprKind::AddrOf(_, r_mut, ref re)) => {
```
Comment (matthewjasper) [inline clippy_lints/src/utils/hir_utils.rs]: This should compare the BorrowKinds

File: clippy_lints/src/utils/hir_utils.rs [hunk 9]
```diff
@@ -398,7 +398,7 @@ impl<'a, 'tcx> SpanlessHash<'a, 'tcx> {
         std::mem::discriminant(&e.kind).hash(&mut self.s);
 
         match e.kind {
-            ExprKind::AddrOf(m, ref e) => {
+            ExprKind::AddrOf(_, m, ref e) => {
```
Comment (matthewjasper) [inline clippy_lints/src/utils/hir_utils.rs]: This should hash the BorrowKind

File: clippy_lints/src/utils/inspector.rs [hunk 10]
```diff
@@ -264,7 +264,7 @@ fn print_expr(cx: &LateContext<'_, '_>, expr: &hir::Expr, indent: usize) {
             println!("{}Relative Path, {:?}", ind, ty);
             println!("{}seg: {:?}", ind, seg);
         },
-        hir::ExprKind::AddrOf(ref muta, ref e) => {
+        hir::ExprKind::AddrOf(_, ref muta, ref e) => {
```
Comment (matthewjasper) [inline clippy_lints/src/utils/inspector.rs]: This should print out the BorrowKind

File: clippy_lints/src/mutable_debug_assertion.rs (hunk 4 above)
Comment (Manishearth) [reply clippy_lints/src/mutable_debug_assertion.rs]: eh, i think this is edge casey enough to leave it focused on Ref.

**Comment by Manishearth** [comment]
Updated, added some more fixes

@bors r+

**Comment by flip1995** [comment]
Thanks! Hadn't had access to my dev machine the last 2 days!

**Pull request merged**